### ISSUE TITLE
question about Low-level interface

### ISSUE THREAD

**Issue by cjfcsjt** [CyberAgent/cmaes#9]
Hi, I have a question about setting bounds in the Low-level interface. I want to set a [-1,1]^100 hypercube bounds, and my code are as below:

```python
    bounds = np.zeros((100,2))
    for i in range(high_dim):
        bounds[i][0] = -1
        bounds[i][1] = 1
    cma_es = CMA(bounds=bounds,mean=np.zeros(high_dim), sigma=1.3)
```

However, this doesn't work when I run "x = cma_es.ask()"
Could you tell me the right way to set the bounds in Low-level interface?

**Comment by c-bata** [comment]
Hi @cjfcsjt ! Thank you for using this package.

It looks like the code you pasted here has no problem. I confirmed it worked.

```python
import numpy as np
from cmaes import CMA


def quadratic(x1, x2):
    return (x1 - 3) ** 2 + (10 * (x2 + 2)) ** 2


def main():
    bounds = np.zeros((2, 2))
    bounds[:, 0] = -1
    bounds[:, 1] = 1

    cma_es = CMA(
        mean=np.zeros(2),
        sigma=1.3,
        bounds=bounds,
        seed=0,
    )

    for generation in range(50):
        solutions = []
        for _ in range(cma_es.population_size):
            x = cma_es.ask()
            value = quadratic(x[0], x[1])
            solutions.append((x, value))
            print(f"#{generation} {value} (x1={x[0]}, x2 = {x[1]})")
        cma_es.tell(solutions)


if __name__ == "__main__":
    main()
```

```
$ python examples/quadratic_function.py 
#0 651.8262344691434 (x1=-0.1341845073316252, x2 = 0.5337780525198841)
#0 469.8157386773097 (x1=0.9893490426910915, x2 = 0.15817752144067695)
#0 598.1976877319871 (x1=0.5770222025690533, x2 = 0.4337766255865469)
#0 317.23240728150915 (x1=0.05948607249187989, x2 = -0.2433390050335837)
#0 628.6446596018732 (x1=0.2014316534059912, x2 = 0.4916112754828256)
#0 497.3509461620192 (x1=-0.4522857941239984, x2 = 0.20325365983517407)
#1 493.21141850186535 (x1=0.6069796344375971, x2 = 0.20790595821440955)
#1 533.8588811864365 (x1=0.7478064437508978, x2 = 0.2995358344057751)
#1 740.7209718853396 (x1=-0.015173555383713844, x2 = 0.7048654316180947)
#1 397.7462571037072 (x1=-0.12629635255539157, x2 = -0.030298174292093263)
#1 361.51241492569625 (x1=-0.012837587707869635, x2 = -0.12267417693200927)
#1 587.4977845814467 (x1=0.6854787434585663, x2 = 0.41275936623291265)
#2 525.9736726492017 (x1=0.2929672373116927, x2 = 0.27737929706699055)
#2 331.89145716428186 (x1=0.2723212972708705, x2 = -0.19874703037313549)
#2 834.5645060311273 (x1=-0.22233448896947183, x2 = 0.8708553890301145)
#2 603.4168455195414 (x1=0.02072952128606641, x2 = 0.4383207191306994)
#2 244.8653202485428 (x1=-0.12909543115429478, x2 = -0.4667874184208112)
#2 149.97663714995673 (x1=0.7948184720750783, x2 = -0.7953680579577203)
#3 237.0490063459266 (x1=0.5207893349820665, x2 = -0.4804523014252301)
#3 499.0106256962565 (x1=0.11316673895577245, x2 = 0.21512261380535158)
#3 500.09248149926816 (x1=-0.016771794854388566, x2 = 0.21583295723535778)
#3 384.1423864364779 (x1=-0.2850058803902758, x2 = -0.06777039976539201)
#3 249.82652863907467 (x1=-0.24753052390746016, x2 = -0.4531319573558836)
#3 202.02429601836548 (x1=-0.08929799271236921, x2 = -0.6126264600671208)
#4 213.33470421529378 (x1=-0.18110476875098352, x2 = -0.5744640423141774)
#4 437.2500997877442 (x1=-0.4292397708756021, x2 = 0.06274189946679798)
#4 241.11539114856808 (x1=0.5546482653934871, x2 = -0.4665866635423279)
#4 277.20647866113 (x1=-0.39711831434364137, x2 = -0.370073419385152)
#4 114.24845311438845 (x1=0.10060259042015798, x2 = -0.9712048416924285)
#4 175.30750523539768 (x1=-0.530861065155233, x2 = -0.7239101701918921)
#5 119.41683257823522 (x1=0.5463634311166685, x2 = -0.9351220719440482)
#5 216.0981423191312 (x1=-0.060924137184231314, x2 = -0.5621930388764487)
#5 210.94606937923368 (x1=-0.45636573956784354, x2 = -0.5893278011757115)
#5 209.84271369005268 (x1=-0.1625175361301812, x2 = -0.5863480041972078)
#5 192.53040087024186 (x1=-0.5255428842681861, x2 = -0.6579830543490599)
#5 117.5508655060181 (x1=0.5440261553572847, x2 = -0.9439741576057478)
#6 276.24704521664495 (x1=0.19232460864456874, x2 = -0.3618180683653539)
#6 111.67296486006876 (x1=0.0793068509119082, x2 = -0.984408958345247)
#6 211.46004998558405 (x1=-0.21368146526075738, x2 = -0.5817887977264333)
#6 253.66659484097153 (x1=0.734816006232114, x2 = -0.42349901263795814)
#6 133.48267870711234 (x1=-0.5700312100556136, x2 = -0.9011935754358678)
#6 190.22094193841943 (x1=0.955909612535434, x2 = -0.6360255265354327)
#7 131.41452172604667 (x1=0.3144575444291258, x2 = -0.8855387694254545)
#7 203.26870529632308 (x1=0.17747756863785003, x2 = -0.602494821402151)
#7 202.49308428233718 (x1=-0.505600881736649, x2 = -0.6208558931703169)
#7 198.53773357933989 (x1=0.7684960645847221, x2 = -0.6087483198013164)
#7 188.9164078274001 (x1=-0.41595140821688537, x2 = -0.6686560031152686)
#7 178.86265919785092 (x1=0.31690816095402574, x2 = -0.6897951405177286)
#8 170.50838905834397 (x1=0.2441691096938608, x2 = -0.7236231545410329)
#8 173.26636731197857 (x1=0.8193595488253133, x2 = -0.7018814594395502)
#8 127.95808291593032 (x1=0.9269711181497352, x2 = -0.8879719690091218)
#8 204.6095919671167 (x1=0.22057736341398315, x2 = -0.5968449772945615)
#8 177.01424520508291 (x1=-0.3884813307445747, x2 = -0.7134058935457601)
#8 213.83060391502525 (x1=0.12933477842162475, x2 = -0.5661594053011136)
#9 122.47586017296108 (x1=0.8206970118740391, x2 = -0.9149815731569055)
#9 109.01203722870801 (x1=0.29688905235505486, x2 = -0.9915098987425308)
#9 160.2112939601902 (x1=0.8468753037666862, x2 = -0.7527015272892361)
#9 131.79114316362714 (x1=0.7792120369233012, x2 = -0.8736819099975319)
#9 113.76483003915615 (x1=0.9795868684403767, x2 = -0.9527046222914112)
#9 188.14851203386414 (x1=0.9309208844797172, x2 = -0.6440227743520838)
```

Can you paste a complete example to reproduce your problem?

**Comment by cjfcsjt** [comment]
Thanks for your response!  @c-bata 
Yeah, when the dimension is 2(or 3), it worked well. But when I set the dimension = 1000, in your code, the "bounds = np.zeros((2, 2)) " changes to "bounds = np.zeros((1000, 2)) ", it seems something wrong.
Here is the exception information from pycharm console:
"Exception: failed to sample a solution in the feasible domain."
I guess that there is a threshold in the dimension parameter.

**Comment by c-bata** [comment]
I got it. It is actually a known issue.

The method for bound constraints is written in B.5 in the paper, https://arxiv.org/abs/1604.00772.
I adopt a re-sampling strategy for bound constraints. But it has an issue like that.

To resolve your issue, I need to implement a repair method, just clipping a value with lower and higher bounds. Could you wait a few days? I'll implement it.

**Comment by c-bata** [comment]
I've just implemented the repair method. Your problem will be fixed after this PR is merged.
https://github.com/c-bata/cmaes/pull/14

I'm going to publish v0.3.0 with this change tomorrow.

**Comment by cjfcsjt** [comment]
Thank you so much!

**Comment by c-bata** [comment]
Hi @cjfcsjt ! I released v0.3.0 which includes the change. Please execute `pip install -U cmaes`.
https://github.com/CyberAgent/cmaes/releases/tag/v0.3.0
